lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+                                                 OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  auto shape = rangeOp.getType().cast<ShapedType>().getShape();
+
+  auto start = rangeOp.getStart();
+  auto end = rangeOp.getEnd();
+  auto stride = (end - start + shape[0] - 1) / shape[0];
+  assert(stride == 1 &&
+         "Expect make_range op to always return tensor of stride 1");
+
+  state.offsets.push_back(builder.getIndexAttr(start));
+  state.sizes.push_back(builder.getIndexAttr(shape[0]));
+  state.strides.push_back(builder.getIndexAttr(stride));
+  state.modulos.push_back(builder.getIndexAttr(0));
+  return success();
+}
+
+LogicalResult
+PtrAnalysis::visitOperandExpandDims(triton::ExpandDimsOp expandDimsOp,
+                                    PtrSState &state, const Location loc,
+                                    OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  if (visitOperand(expandDimsOp.getSrc(), state, loc, builder).failed()) {
+    return failure();
+  }
+
+  auto dstShape =
+      expandDimsOp.getResult().getType().cast<ShapedType>().getShape();
+  auto axis = expandDimsOp.getAxis();
+
+  assert(dstShape[axis] == 1 &&
+         "expect changed dimension to be 1 in expand_dims");
+
+  // insert dimension info
+  state.offsets.insert(state.offsets.begin() + axis, builder.getIndexAttr(0));
+  state.sizes.insert(state.sizes.begin() + axis, builder.getIndexAttr(1));
+  state.strides.insert(state.strides.begin() + axis, builder.getIndexAttr(0));
+  state.modulos.insert(state.modulos.begin() + axis, builder.getIndexAttr(0));
+
+  if (state.hasModulo() && state.getRank() > 2) {
+    expandDimsOp->emitRemark(
+        "PtrAnalysis: unsupported scenario where expand_dims result "
+        "has modulo and rank > 2");
+    return failure();
+  }
+
+  return success();
+}
+
+LogicalResult
+PtrAnalysis::visitOperandBroadcast(triton::BroadcastOp broadcastOp,
+                                   PtrSState &state, const Location loc,
+                                   OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  auto src = broadcastOp.getSrc();
+  auto dst = broadcastOp.getResult();
+
+  SmallVector<int64_t> srcShape;
+  auto dstShape = dst.getType().cast<ShapedType>().getShape();
+
+  if (src.getType().isa<ShapedType>()) {
+    srcShape =
+        SmallVector<int64_t>(src.getType().cast<ShapedType>().getShape());
+    assert(srcShape.size() == dstShape.size() &&
+           "rank of source and destination should match");
+
+    if (visitOperand(src, state, loc, builder).failed()) {
+      return failure();
+    }
+
+    for (size_t i = 0; i < dstShape.size(); i++) {
+      if (srcShape[i] == dstShape[i]) {
+        continue;
+      } else if (srcShape[i] < dstShape[i]) {
+        state.sizes[i] = builder.getIndexAttr(dstShape[i]);
+      } else {
+        llvm_unreachable("unexpected dimensions used in broadcast");
+      }
+    }
+    return success();
+  }
+
```

Comment:
nit:
- this code looks quite nested; i think we can reverse the condition `src.getType().isa<ShapedType>()` and fail early
- `srcShape` declaration and assignment should be on the same line, here it's split but the initial value isn't used elsewhere
- should probably change to `cast<ShapedType>(src.getType())` similar to my other comment since this style cast isn't preferred now

Reply:
Thanks for the comment. This code is much cleaner now.